src/Catalyst.Core.Modules.Dfs/CoreApi/FileSystemApi.cs

```diff
@@ -0,0 +1,404 @@
+using Lib.P2P;
+using MultiFormats;
+using ProtoBuf;
+
+namespace Catalyst.Core.Modules.Dfs.CoreApi
+{
+    internal class FileSystemApi : IFileSystemApi
+    {
+        private static ILog log = LogManager.GetLogger(typeof(FileSystemApi));
+        private IpfsEngine ipfs;
+
+        private static readonly int DefaultLinksPerBlock = 174;
+
+        public FileSystemApi(IpfsEngine ipfs) { this.ipfs = ipfs; }
+
+        public async Task<IFileSystemNode> AddFileAsync(string path,
+            AddFileOptions options = default(AddFileOptions),
+            CancellationToken cancel = default(CancellationToken))
+        {
+            using (var stream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read))
+            {
+                return await AddAsync(stream, Path.GetFileName(path), options, cancel).ConfigureAwait(false);
+            }
+        }
+
+        public async Task<IFileSystemNode> AddTextAsync(string text,
+            AddFileOptions options = default(AddFileOptions),
+            CancellationToken cancel = default(CancellationToken))
+        {
+            using (var ms = new MemoryStream(Encoding.UTF8.GetBytes(text), false))
+            {
+                return await AddAsync(ms, "", options, cancel).ConfigureAwait(false);
+            }
+        }
+
+        public async Task<IFileSystemNode> AddAsync(Stream stream,
+            string name,
+            AddFileOptions options,
+            CancellationToken cancel)
+        {
+            options = options ?? new AddFileOptions();
+
+            // TODO: various options
+            if (options.Trickle) throw new NotImplementedException("Trickle");
+
+            var blockService = GetBlockService(options);
+            var keyChain = await ipfs.KeyChainAsync(cancel).ConfigureAwait(false);
+
+            var chunker = new SizeChunker();
+            var nodes = await chunker.ChunkAsync(stream, name, options, blockService, keyChain, cancel)
+               .ConfigureAwait(false);
+
+            // Multiple nodes for the file?
+            var node = await BuildTreeAsync(nodes, options, cancel);
+
+            // Wrap in directory?
+            if (options.Wrap)
+            {
+                var link = node.ToLink(name);
+                var wlinks = new IFileSystemLink[] {link};
+                node = await CreateDirectoryAsync(wlinks, options, cancel).ConfigureAwait(false);
+            }
+            else
+            {
+                node.Name = name;
+            }
+
+            // Advertise the root node.
+            if (options.Pin && ipfs.IsStarted) await ipfs.Dht.ProvideAsync(node.Id, true, cancel).ConfigureAwait(false);
+
+            // Return the file system node.
+            return node;
+        }
+
+        private async Task<FileSystemNode> BuildTreeAsync(IEnumerable<FileSystemNode> nodes,
+            AddFileOptions options,
+            CancellationToken cancel)
+        {
+            if (nodes.Count() == 1) return nodes.First();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)